sw/device/silicon_creator/lib/drivers/kmac.c

```diff
@@ -0,0 +1,279 @@
+ *
+ * Sets the KMAC block to use software entropy with an all-zero seed (since we
+ * have no secret inputs for SPHINCS+) and sets the mode to SHAKE-256.
+ *
+ * @return Error code indicating if the operation succeeded.
+ */
+rom_error_t kmac_shake256_configure(void) {
+  HARDENED_RETURN_IF_ERROR(poll_state(KMAC_STATUS_SHA3_IDLE_BIT));
+
+  uint32_t entropy_period_reg = KMAC_ENTROPY_PERIOD_REG_RESVAL;
+  // Set the wait timer to the maximum count.
+  entropy_period_reg = bitfield_field32_write(
+      entropy_period_reg, KMAC_ENTROPY_PERIOD_WAIT_TIMER_FIELD,
+      KMAC_ENTROPY_PERIOD_WAIT_TIMER_MASK);
+  // Set the prescaler to the maximum number of cycles.
+  entropy_period_reg = bitfield_field32_write(
+      entropy_period_reg, KMAC_ENTROPY_PERIOD_PRESCALER_FIELD,
+      KMAC_ENTROPY_PERIOD_PRESCALER_MASK);
+  abs_mmio_write32(kBase + KMAC_ENTROPY_PERIOD_REG_OFFSET, entropy_period_reg);
+
+  uint32_t cfg_reg = KMAC_CFG_SHADOWED_REG_RESVAL;
+  // Set `CFG.KMAC_EN` bit to 0.
+  // NOTE: If this driver is ever modified to perform an operation with
+  // KMAC_EN=true and use entropy from EDN, then the absorb() function must
+  // poll `STATUS.fifo_depth` to avoid a specific EDN-KMAC-Ibex deadlock
+  // scenario. See `absorb()` and the KMAC documentation for details.
+  cfg_reg = bitfield_bit32_write(cfg_reg, KMAC_CFG_SHADOWED_KMAC_EN_BIT, 0);
+  // Set `CFG.KSTRENGTH` field to 256-bit strength.
+  cfg_reg = bitfield_field32_write(cfg_reg, KMAC_CFG_SHADOWED_KSTRENGTH_FIELD,
+                                   KMAC_CFG_SHADOWED_KSTRENGTH_VALUE_L256);
+  // Set `CFG.MODE` field to SHAKE.
+  cfg_reg = bitfield_field32_write(cfg_reg, KMAC_CFG_SHADOWED_MODE_FIELD,
+                                   KMAC_CFG_SHADOWED_MODE_VALUE_SHAKE);
+  // Set `CFG.MSG_ENDIANNESS` bit to 0 (little-endian).
+  cfg_reg =
+      bitfield_bit32_write(cfg_reg, KMAC_CFG_SHADOWED_MSG_ENDIANNESS_BIT, 0);
+  // Set `CFG.STATE_ENDIANNESS` bit to 0 (little-endian).
+  cfg_reg =
+      bitfield_bit32_write(cfg_reg, KMAC_CFG_SHADOWED_STATE_ENDIANNESS_BIT, 0);
+  // Set `CFG.SIDELOAD` bit to 0 (no sideloading).
+  cfg_reg = bitfield_bit32_write(cfg_reg, KMAC_CFG_SHADOWED_SIDELOAD_BIT, 0);
+  // Set `CFG.ENTROPY_MODE` field to use software entropy. SHAKE does not
+  // require any entropy, so there is no reason we should wait for entropy
+  // availability before we start hashing.
+  cfg_reg =
+      bitfield_field32_write(cfg_reg, KMAC_CFG_SHADOWED_ENTROPY_MODE_FIELD,
+                             KMAC_CFG_SHADOWED_ENTROPY_MODE_VALUE_SW_MODE);
+  // Set `CFG.ENTROPY_FAST_PROCESS` bit to 0.
+  cfg_reg = bitfield_bit32_write(cfg_reg,
+                                 KMAC_CFG_SHADOWED_ENTROPY_FAST_PROCESS_BIT, 0);
+  // Set `CFG.MSG_MASK` bit to 0.
+  cfg_reg = bitfield_bit32_write(cfg_reg, KMAC_CFG_SHADOWED_MSG_MASK_BIT, 0);
+  // Set `CFG.ENTROPY_READY` bit to 1.
+  cfg_reg =
+      bitfield_bit32_write(cfg_reg, KMAC_CFG_SHADOWED_ENTROPY_READY_BIT, 1);
```

Comment:
Technically no, the entropy wait timer shouldn't matter because the entropy mode is SW. It's also true that we don't expect KMAC to wait for entropy for SHAKE under normal circumstances. However, if the entropy mode is EDN, KMAC waits to get initial entropy before it starts any process, which means a long delay.